### PR TITLE
[SPARK-14180][Core]Fix a deadlock in CoarseGrainedExecutorBackend Shutdown

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

Call `executor.stop` in a new thread to eliminate deadlock.
## How was this patch tested?

Existing unit tests

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala [hunk 1]
```diff
@@ -113,9 +113,9 @@ private[spark] class CoarseGrainedExecutorBackend(
 
     case Shutdown =>
       stopping.set(true)
-      executor.stop()
-      stop()
-      rpcEnv.shutdown()
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: `rpcEnv.shutdown()` will be called in `executor.stop()` and `endpoint.stop()` will be called by `rpcEnv.shutdown()`, so we don't need these two lines.

**Comment by mgummelt** [comment]
Can you write a regression test?  This bug wasn't caught by existing tests.

File: core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala [hunk 2]
```diff
@@ -113,9 +113,9 @@ private[spark] class CoarseGrainedExecutorBackend(
...
+      new Thread("CoarseGrainedExecutorBackend-stop-executor") {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: can you add a comment on why we need to do this

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala]: Added

**Comment by zsxwing** [comment]
> Can you write a regression test? This bug wasn't caught by existing tests.

I tried. But it's really hard to write such test. E.g., even if this dead lock happens, the executor could still be killed by other mechanism.

**Comment by andrewor14** [comment]
LGTM merging into master

**Pull request closed without merging**